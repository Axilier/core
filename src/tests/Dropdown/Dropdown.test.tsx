import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { Dropdown, Option } from '../../index';
import { expectToThrow } from '../expectToThrow';

const wrapper = document.createElement('div');
wrapper.id = 'test-wrapper';
document.body.appendChild(wrapper);
const onChange = jest.fn();
const dropdown = mount(
    <Dropdown label={'test'} onChange={onChange}>
        <Option>test1</Option>
        <Option>test2</Option>
    </Dropdown>,
    { attachTo: wrapper },
);

describe('when the dropdown is used in an application', () => {
    it('should render the contained dropdown correctly', () => {
        expect(dropdown).toMatchSnapshot();
    });
    describe('when I render the component with children that arent Options', () => {
        it('should throw an error', () => {
            expectToThrow(
                () =>
                    mount(
                        <Dropdown>
                            <div />
                        </Dropdown>,
                    ),
                'Children of Dropdown must be of type Option',
            );
        });
    });
    describe('when the mainBody of the dropdown is clicked', () => {
        beforeEach(() => dropdown.find('.mainBody').simulate('click'));
        it('should open the dropdown with the listed options', () => {
            expect(dropdown.find('.openMenu').props().style).toHaveProperty(
                'visibility',
                'visible',
            );
        });
        it('should close the dropdown with the listed options', () => {
            expect(dropdown.find('.openMenu').props().style).toHaveProperty(
                'visibility',
                'hidden',
            );
        });
    });
    describe('when the dropdown is open', () => {
        beforeAll(() => {
            dropdown.find('.mainBody').simulate('click');
        });
        it('should show the dropdown open', () => {
            expect(dropdown.find('.openMenu').props().style).toHaveProperty(
                'visibility',
                'visible',
            );
        });
        describe('when the mainBody is clicked', () => {
            beforeAll(() => {
                act(() => {
                    document.getElementById('DropdownMainBody')?.dispatchEvent(
                        new MouseEvent('click', {
                            bubbles: true,
                        }),
                    );
                });
                dropdown.update();
            });
            it('should do nothing', () => {
                expect(dropdown.find('.openMenu').props().style).toHaveProperty(
                    'visibility',
                    'hidden',
                );
            });
        });
        describe('when any region of the screen other than mainBody is clicked', () => {
            beforeAll(() => {
                act(() => {
                    document.dispatchEvent(
                        new MouseEvent('click', {
                            clientX: 10,
                            clientY: 10,
                            bubbles: true,
                        }),
                    );
                });
                dropdown.update();
            });
            it('should close the dropdown', () => {
                expect(dropdown.find('.openMenu').props().style).toHaveProperty(
                    'visibility',
                    'hidden',
                );
            });
        });
        describe('when any other than enter key is pressed', () => {
            beforeAll(() =>
                dropdown.find('.mainBody').simulate('keydown', { key: 'a' }),
            );
            it('should do nothing', () => {
                expect(dropdown.find('.openMenu').props().style).toHaveProperty(
                    'visibility',
                    'hidden',
                );
            });
        });
        describe('when the enter key is pressed', () => {
            beforeAll(() =>
                dropdown
                    .find('.mainBody')
                    .simulate('keydown', { key: 'Enter' }),
            );
            it('should toggle the status of the menu', () => {
                expect(dropdown.find('.openMenu').props().style).toHaveProperty(
                    'visibility',
                    'visible',
                );
            });
        });
        describe('when one of the child Options is selected', () => {
            beforeEach(() =>
                dropdown.find('.openMenu').childAt(1).simulate('click'),
            );
            it('should close the menu', () => {
                expect(dropdown.find('.openMenu').props().style).toHaveProperty(
                    'visibility',
                    'hidden',
                );
                expect(dropdown.find('.mainBody').childAt(0).text()).toBe(
                    'test2',
                );
            });
            describe('when `onChange` is not defined', () => {
                beforeAll(() => dropdown.setProps({ onChange: null }));
                it('should do nothing', () => {
                    expect(true).toBeTruthy();
                });
            });
        });
    });
    describe('when the `required` prop is false', () => {
        it('shouldn`t render a * with the label', () => {
            expect(dropdown.find('.required').exists()).toBeFalsy();
        });
    });
    describe('when the `required` prop is true', () => {
        beforeAll(() => dropdown.setProps({ required: true }));
        it('should render a * with the label', () => {
            expect(dropdown.find('.required').exists()).toBeTruthy();
        });
    });
    describe('when the `label` is not defined', () => {
        beforeAll(() => dropdown.setProps({ label: undefined }));
        it('should not render the label', () => {
            expect(dropdown.find('.label').exists()).toBeFalsy();
        });
    });
});
