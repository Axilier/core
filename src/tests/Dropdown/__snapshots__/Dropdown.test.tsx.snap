// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the dropdown is used in an application should render the contained dropdown correctly 1`] = `
<Dropdown
  label="test"
  onChange={[MockFunction]}
>
  <div
    className="dropdown undefined"
  >
    <div
      className="label"
    >
      test
    </div>
    <div
      className="mainBody"
      id="DropdownMainBody"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      style={
        Object {
          "height": "45px",
        }
      }
    >
      <Option>
        <div
          className="option"
          onClick={[Function]}
          role="button"
        >
          test1
        </div>
      </Option>
      <svg
        className="dropArrow"
        fill="black"
        height="16"
        viewBox="0 0 16 16"
        width="16"
      >
        <path
          d="M7.247 11.14 2.451 5.658C1.885 5.013 2.345 4 3.204 4h9.592a1 1 0 0 1 .753 1.659l-4.796 5.48a1 1 0 0 1-1.506 0z"
        />
      </svg>
    </div>
    <div
      className="openMenu"
      id="openMenu"
      style={
        Object {
          "visibility": "hidden",
        }
      }
    >
      <Option
        key="test-0"
        onClick={[Function]}
      >
        <div
          className="option"
          onClick={[Function]}
          role="button"
        >
          test1
        </div>
      </Option>
      <Option
        key="test-1"
        onClick={[Function]}
      >
        <div
          className="option"
          onClick={[Function]}
          role="button"
        >
          test2
        </div>
      </Option>
    </div>
  </div>
</Dropdown>
`;
